From 4a73c45f4c57cbc8e30e5ca15e90887f04050981 Mon Sep 17 00:00:00 2001
From: Megha Aggarwal <megha.aggarwal@intel.com>
Date: Fri, 24 Jul 2015 21:58:09 +0530
Subject: [PATCH] FOR_UPSTREAM [VPG]: drm/i915: Fix for hang issues in Command
 mode.

As there are mutliple mem_write request in a single atmoic path
that will keep the pipe busy and getting Flip error timeoue.
To fix this issue removed multiple mem_write() call to
single mem_write() in the atmoic path.

Change-Id: I14d07eb089bfd482180355c0ec3b7dafc076c35f
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-7482
Signed-off-by: Megha Aggarwal <megha.aggarwal@intel.com>
Signed-off-by: Ramalingam C <ramalingam.c@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c | 3 ++-
 drivers/gpu/drm/i915/intel_dsi.c     | 3 +++
 drivers/gpu/drm/i915/intel_dsi.h     | 2 ++
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 0c68911..7db2b7a 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11009,7 +11009,6 @@ static void i915_commit(struct drm_i915_private *dev_priv,
 		POSTING_READ(DSPCNTR(pipe));
 	}
 
-	intel_dsi_send_fb_on_crtc(&intel_crtc->base);
 	reg->surf = 0;
 }
 
@@ -11459,6 +11458,8 @@ static int intel_crtc_set_display(struct drm_crtc *crtc,
 
 	intel_crtc->atomic_update = false;
 
+	intel_dsi_send_fb_on_crtc(&intel_crtc->base);
+
 	intel_runtime_pm_put(dev_priv);
 	return ret;
 }
diff --git a/drivers/gpu/drm/i915/intel_dsi.c b/drivers/gpu/drm/i915/intel_dsi.c
index cc41178..4cd869d 100644
--- a/drivers/gpu/drm/i915/intel_dsi.c
+++ b/drivers/gpu/drm/i915/intel_dsi.c
@@ -1127,6 +1127,9 @@ static void intel_dsi_mode_set(struct intel_encoder *intel_encoder)
 					IP_TG_CONFIG |
 					RANDOM_DPI_DISPLAY_RESOLUTION);
 
+		if (is_cmd_mode(intel_dsi))
+			I915_WRITE(MIPI_TEARING_CTRL(pipe),
+						CHV_CMD_MODE_TEARING_DELAY);
 
 		/* For Port C for dual link */
 		if (intel_dsi->dual_link)
diff --git a/drivers/gpu/drm/i915/intel_dsi.h b/drivers/gpu/drm/i915/intel_dsi.h
index 96bdac2..95d341d 100644
--- a/drivers/gpu/drm/i915/intel_dsi.h
+++ b/drivers/gpu/drm/i915/intel_dsi.h
@@ -403,6 +403,8 @@
 #define FLIS_DSI_TXCNTRL		0x15
 #define BIT_FILS_DSI_TXCNTRL_HS_IO_SEL	(1 << 0)
 
+#define CHV_CMD_MODE_TEARING_DELAY	0x20
+
 struct intel_dsi_device {
 	unsigned int panel_id;
 	const char *name;
-- 
1.9.1

